aospy/automate.py

```diff
@@ -0,0 +1,263 @@
+            return requested
+
+    def _permute_core_specs(self):
+        """Generate all requested combinations of the core objects."""
+        obj_trees = []
+        projects = self._get_requested_spec(self._obj_lib, _PROJECTS_STR)
+        for project in projects:
+            models = self._get_requested_spec(project, _MODELS_STR)
+            for model in models:
+                runs = self._get_requested_spec(model, _RUNS_STR)
+                for run in runs:
+                    obj_trees.append({
+                        self._NAMES_SUITE_TO_CALC[_PROJECTS_STR]: project,
+                        self._NAMES_SUITE_TO_CALC[_MODELS_STR]: model,
+                        self._NAMES_SUITE_TO_CALC[_RUNS_STR]: run,
+                    })
+        return obj_trees
+
+    def _get_regions(self):
+        """Get the requested regions."""
+        if self._specs_in[_REGIONS_STR] == 'all':
+            try:
+                return [_get_all_objs_of_type(Region, self._obj_lib.regions)]
+            except AttributeError:
+                return [_get_all_objs_of_type(Region, self._obj_lib)]
+        else:
+            return set(self._specs_in[_REGIONS_STR])
+
+    def _get_variables(self):
+        """Get the requested variables."""
+        if self._specs_in[_VARIABLES_STR] == 'all':
+            try:
+                return _get_all_objs_of_type(Var, self._obj_lib.variables)
+            except AttributeError:
+                return _get_all_objs_of_type(Var, self._obj_lib)
+        else:
+            return set(self._specs_in[_VARIABLES_STR])
+
+    def _get_aux_specs(self):
+        """Get and pre-process all of the non-core specifications."""
+        # Drop the "core" specifications, which are handled separately.
+        specs = self._specs_in.copy()
+        [specs.pop(core) for core in self._CORE_SPEC_NAMES]
+
+        specs[_REGIONS_STR] = self._get_regions()
+        specs[_VARIABLES_STR] = self._get_variables()
+
+        return specs
+
+    def _permute_aux_specs(self):
+        """Generate all permutations of the non-core specifications."""
+        # Convert to attr names that Calc is expecting.
+        calc_aux_mapping = self._NAMES_SUITE_TO_CALC.copy()
+        # Special case: manually make library an iterable
+        calc_aux_mapping[_OBJ_LIB_STR] = [self._obj_lib]
+        [calc_aux_mapping.pop(core) for core in self._CORE_SPEC_NAMES]
+
+        specs = self._get_aux_specs()
+        for suite_name, calc_name in calc_aux_mapping.items():
+            specs[calc_name] = specs.pop(suite_name)
+        return _permuted_dicts_of_specs(specs)
+
+    def _combine_core_aux_specs(self):
+        """Combine permutations over core and auxilliary Calc specs."""
+        all_specs = []
+        for core_dict in self._permute_core_specs():
+            for aux_dict in self._permute_aux_specs():
+                all_specs.append(_merge_dicts(core_dict, aux_dict))
+                print(_merge_dicts(core_dict, aux_dict), '\n')
```

Comment:
For now, in lieu of a nicer repr, could we use [```pprint.pprint```](https://docs.python.org/3/library/pprint.html#pprint.pprint) from the standard library here (and elsewhere) for nicer printing of dicts?

E.g. ```pprint.pprint(_merge_dicts(core_dict, aux_dict))``` produces:
```
{'date_range': (datetime.datetime(3, 1, 1, 0, 0),
                datetime.datetime(6, 12, 31, 0, 0)),
 'dtype_in_time': 'ts',
 'dtype_in_vert': 'sigma',
 'dtype_out_time': ('av', 'reg.av'),
 'dtype_out_vert': None,
 'intvl_in': 'monthly',
 'intvl_out': 'ann',
 'model': Model instance "idealized_moist_rad",
 'proj': Project instance "itcz",
 'region': (Geographical region "globe",),
 'run': aospy.Run instance "control",
 'time_offset': {'days': -15},
 'var': Var instance "precip_im"}
```